---
title: "Parsing the results of gene expression and selection scans."
author: "Luciano V Cosme"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: breezedark
    css:
      - "styles.css"
    toc: yes
    toc_float: no
    toc_depth: 5
editor_options:
  markdown:
    wrap: 120
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  eval                        = TRUE,
  echo                        = TRUE,
  cache                       = TRUE, # tidy = TRUE,
  class.output                = "bg-success"
)
knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```



## 1. R libraries

```{r libraries, message=FALSE, results='hide'}
library(tidyverse)
library(here)
library(dplyr)
library(ggplot2)
library(colorout)
library(extrafont)
library(reticulate)
library(scales)
library(stringr)
library(grid)
library(flextable)
library(devtools)
library(readr)
library(purrr)
library(ggtext)
library(ggvenn)
library(qqplotr)
library(qqman)
library(betareg)
```

## 2. Get gene information from gff file

Since we created a chromosomal scale and the gff file has the scaffolds positions, we can get the genomic coordinates of the genes and then lift them to the chromosomal scale. Another way would be get the SNPs IDs and check where they are located. Each way has its up and down sides.

The gff file with the annotation for the AalbF3 genome assembly is available in the project, under the files directory.

```{bash}
head data/files/genes.gff
```

To start we can get the scaffold, start and end of each gene. I will do this because I might need to represent the genes in a plot using ggplot.

```{bash}
# In the code below we get the gene id, the scaffold id, the gene start, the gene end, and the size of the genomic regions the gene covers (end - start)
awk 'BEGIN {OFS="\t"} $3=="gene" {split($9, a, ";"); split(a[1], b, "="); size = $5 - $4; print b[2], $1, $4, $5, size}' data/files/genes.gff | head
```


Create a file with the genomic coordinates for each gene
```{bash}
# In the code below we get the gene id, the scaffold id, the gene start, the gene end, and the size of the genomic regions the gene covers (end - start)
awk 'BEGIN {OFS="\t"} $3=="gene" {split($9, a, ";"); split(a[1], b, "="); size = $5 - $4; print b[2], $1, $4, $5, size}' data/files/genes.gff > data/files/genes.txt;
wc -l data/files/genes.txt
```

We got 20,621 genes in the gff file.

We can import the data into R now
```{r}
gene_data <- read_delim(here("data", "files","genes.txt"), delim = "\t", col_names = c("Gene_ID", "Scaffold", "Start", "End", "Size"), show_col_types = FALSE)
head(gene_data)
```

First we need to remove the string "chr" from the collumn Scaffold.

```{r}
gene_data <- gene_data %>%
  mutate(Scaffold = str_remove(Scaffold, "chr"))
head(gene_data)
```

Now we can import the bim file in which we have the Scaffold

```{r}
# Import the function
source(
  here(
    "scripts", "analysis", "import_bim.R")
)

# Import the data
scaffolds <- import_bim(here("output", "quality_control", "file1.bim"))
head(scaffolds)
```

Now import the file with chromosomal scale
```{r}
chromosomes <- import_bim(here("output", "quality_control", "file7.bim"))

# Rename columns
chromosomes <- chromosomes |>
  dplyr::rename(Chromosome = Scaffold,
                Position_chr = Position)

head(chromosomes)
```

Now merge the object
```{r}
# Merge and keep only the rows with data
snps_albo <- inner_join(scaffolds, chromosomes, by = "SNP", suffix = c("", "")) |>
  dplyr::select(
    SNP, Chromosome, Position_chr, Scaffold, Position
  )
head(snps_albo)
```

Now we can add our gene data with the scaffold data

```{r, warning=FALSE}
# Perform a non-equi join
merged_scaff <- scaffolds |>
  left_join(gene_data, by = "Scaffold") |>
  filter(Position >= Start & Position <= End)
head(merged_scaff)
```


Now we merge with chromosomal data

```{r}
# Merge
snps_genes_chr <- inner_join(merged_scaff, chromosomes, by = "SNP", suffix = c("", "")) |>
  dplyr::select(
    SNP, Chromosome, Position_chr, Scaffold, Position, Gene_ID, Start, End, End
  ) |>
  mutate(Gene_ID = str_remove(Gene_ID, "gene-"))
head(snps_genes_chr)
```

Save it
```{r}
saveRDS(snps_genes_chr, file = here("output", "ldna", "snps_genes_chr.rds"))
```


## 3. Expression data
Now we can import gene expression data

```{r}
gene_expression <- read_delim(here("data", "files","MANvsAUTO_sig_mRNAs.csv"), delim = ",", col_names = TRUE, show_col_types = FALSE) |>
  dplyr::select(
    gene,log2FoldChange 
  ) |>
  dplyr::rename(
    Gene_ID = gene
  )
head(gene_expression)
```

Now we can merge it with our snps_genes_chr object

```{r}
snps_expression <- inner_join(snps_genes_chr, gene_expression, by = "Gene_ID", suffix = c("", ""))
head(snps_expression)
```

Check how many SNPs we have in the genes with differential expression
```{r}
length(snps_expression$SNP)
```

Save the list of SNPs on DE genes
```{r}
# Select only the desired columns
data_to_save <- snps_expression[, c("SNP", "Chromosome", "Position_chr")]

# Write the selected data to a file
write.table(
  data_to_save,
  file = here("output", "snpeff", "SNPs_on_DE_genes.txt"),
  row.names = FALSE,
  quote = FALSE,
  col.names = TRUE,
  sep = "\t" # or use "," for CSV format
)
```


Now we can check if any of the SNPs we found in our analysis are within these genes. If we do not find any, them we can develop a different approach, which is find the nearest gene for the SNPs or the nearest SNP for the genes of interest.

First lets check the 17 SNPs that we find in common

```{r}
snps01<- read_delim(here("output", "pcadapt", "4_way_venn_common_SNPs_pcadapt_outflank.txt"), delim = "\t", col_names = c("SNP"), show_col_types = FALSE)
head(snps01)
```

Check how many SNPs
```{r}
length(snps01$SNP)
```

Check if there is any differentially expressed gene with one of these SNPs
```{r}
# Filter snps_expression to keep only the rows that have a matching SNP in snps01
filtered_snps_expression <- snps_expression |>
  semi_join(snps01, by = "SNP")
head(filtered_snps_expression)
```
None of the 17 SNPs are present in the genes. Now we can expand our search. 

```{r}
# Lets check all the SNPs from the comparisons of the populations vs the autogenous line
# Read data from txt files
MAN_AUT <-
  read.table(
    here("output", "outflank", "man_aut_SNPs_outFlank.txt"),
    stringsAsFactors = FALSE
  ) |>
  drop_na()


NEW_AUT <-
  read.table(
    here("output", "outflank", "new_aut_SNPs_outFlank.txt"),
    stringsAsFactors = FALSE
  )

NEW_MAN_AUT <-
  read.table(
    here("output", "outflank", "SNPs_outFlank.txt"),
    stringsAsFactors = FALSE
    )
```

Merge them and get a SNP count
```{r}
# Combine all three data frames into one
combined_df <- bind_rows(MAN_AUT, NEW_AUT, NEW_MAN_AUT)

# Count unique values and add a 'Count' column
result <- combined_df %>%
  group_by(V1) %>%
  summarise(Count = n()) %>%
  arrange(V1) |>
  dplyr::rename(SNP = V1)

# Show the result
head(result)
```

How many SNPs?
```{r}
length(unique(result$SNP))
```

How many SNPs out of the 158 are on genes with differencial expression
```{r}
filtered_snps_expression2 <- snps_expression |>
  semi_join(result, by = "SNP")
length(filtered_snps_expression2$SNP)
```

```{r}
unique(filtered_snps_expression2$SNP)
```

Save it
```{r}
write.table(
  filtered_snps_expression2$SNP,
  file = here("output", "snpeff","SNPs_79_DE.txt"),
  row.names = FALSE,
  quote = FALSE,
  col.names = FALSE,
  sep = "\n"
)
```


We can also check if any of the 157 outliers

Select the 17 SNPs outliers identified in all 3 comparisons where we had the AUT line. See the 4 way Venn diagram
```{r}
snps_17 <-
  read.table(
    here("output", "pcadapt", "4_way_venn_common_SNPs_pcadapt_outflank.txt"),
    stringsAsFactors = FALSE
    )

# Get the 17 SNPs 
snps_aut17 <- snps_expression |> 
  filter(SNP %in% snps_17$V1)

head(snps_aut17)
```



Lets look what are the nearest gene of all the 158 SNPs, then we can filter out whatever way we want. We need to do SNP functional annotation for all the SNPs.

To do the functional annotation we need to include all SNPs, so we can use file1.

Let's create a vcf file for the annotation

```{bash}
plink2 \
--allow-extra-chr \
--bfile output/quality_control/file1 \
--recode vcf \
--out output/snpeff/albo \
--silent;
grep "samples\|variants" output/snpeff/albo.log
```

Now check the R Markdown with functional annotation

Create bed file

```{r}
snps_functional <- snps_genes_chr %>%
  filter(SNP %in% result$SNP) |>
  dplyr::select(
    SNP, Scaffold, Position
  )
head(snps_functional)
```

```{r}
# Assuming snps_functional is your tibble
write.table(snps_functional %>% 
              select(SNP, Scaffold, Position) %>% 
              mutate(End = Position + 1),
            file = here("output", "snpeff","snps_functional.bed"),
            quote = FALSE,
            row.names = FALSE,
            col.names = FALSE,
            sep = "\t")

```

Check it out
```{bash}
head output/snpeff/snps_functional.bed
```





